File: apps/backend-node/src/features/profile/profile.service.ts
import { Injectable } from '@nestjs/common';

import type { AuthUser } from '../../shared/auth/auth-user.model';
import { NotFoundError } from '../../shared/errors';
import type { ProfileDto, UpdateProfileDto } from './dto';
import { mapProfile } from './mappers/profile.mapper';
import { ProfileRepository } from './profile.repository';

@Injectable()
export class ProfileService {
  constructor(private readonly repository: ProfileRepository) {}

  async getProfile(user: AuthUser): Promise<ProfileDto> {
    const profile = await this.repository.getById(user.id);
    if (!profile) {
      throw new NotFoundError('Profile not found');
    }

    return mapProfile(profile);
  }

  async updateProfile(
    user: AuthUser,
    request: UpdateProfileDto,
  ): Promise<ProfileDto> {
    const phone = request.phone?.trim() ? request.phone.trim() : null;
    const avatarUrl = request.avatarUrl?.trim()
      ? request.avatarUrl.trim()
      : null;

    const updated = await this.repository.update(user.id, {
      fullName: request.fullName,
      phone,
      preferredLanguage: request.preferredLanguage,
      marketingConsent: request.marketingConsent,
      avatarUrl,
    });

    if (!updated) {
      throw new NotFoundError('Profile not found');
    }

    return mapProfile(updated);
  }
}

File: apps/backend-java/src/main/java/com/housingplatform/profile/ProfileService.java
package com.housingplatform.profile;

import com.housingplatform.auth.error.NotFoundException;
import com.housingplatform.auth.model.AuthenticatedUser;
import com.housingplatform.profile.dto.ProfileDto;
import com.housingplatform.profile.dto.UpdateProfileRequest;
import com.housingplatform.profile.mapper.ProfileMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ProfileService {

  private final ProfileRepository profileRepository;

  public ProfileService(ProfileRepository profileRepository) {
    this.profileRepository = profileRepository;
  }

  @Transactional(readOnly = true)
  public ProfileDto getProfile(AuthenticatedUser user) {
    return profileRepository
        .findById(user.id())
        .map(ProfileMapper::toDto)
        .orElseThrow(() -> new NotFoundException("Profile not found"));
  }

  @Transactional
  public ProfileDto updateProfile(AuthenticatedUser user, UpdateProfileRequest request) {
    String phone = request.phone() == null || request.phone().isBlank() ? null : request.phone().strip();
    String avatarUrl =
        request.avatarUrl() == null || request.avatarUrl().isBlank()
            ? null
            : request.avatarUrl().strip();
    var updated =
        profileRepository.update(
            user.id(),
            request.fullName().strip(),
            phone,
            request.preferredLanguage().strip(),
            request.marketingConsent(),
            avatarUrl);
    if (updated == null) {
      throw new NotFoundException("Profile not found");
    }
    return ProfileMapper.toDto(updated);
  }
}
